subxt/src/rpc/lightclient/background.rs

```diff
@@ -0,0 +1,275 @@
+                    }
+                };
+            }
+            Err(err) => {
+                tracing::warn!(target: LOG_TARGET, "annot decode RPC response {:?}", err);
+            }
+        }
+    }
+
+    /// Perform the main background task:
+    /// - receiving user registration for RPC method / subscriptions
+    /// - providing the results from the light client back to users.
+    pub async fn start_task(
+        &mut self,
+        backend: mpsc::Receiver<BackendMessage>,
+        rpc_responses: smoldot_light::JsonRpcResponses,
+    ) {
+        let backend_event = tokio_stream::wrappers::ReceiverStream::new(backend);
+        let rpc_responses_event =
+            futures_util::stream::unfold(rpc_responses, |mut rpc_responses| async {
+                rpc_responses
+                    .next()
+                    .await
+                    .map(|result| (result, rpc_responses))
+            });
+
+        tokio::pin!(backend_event, rpc_responses_event);
+
+        let mut backend_event_fut = backend_event.next();
+        let mut rpc_responses_fut = rpc_responses_event.next();
+
+        loop {
+            match future::select(backend_event_fut, rpc_responses_fut).await {
+                // Message received from the backend: user registered.
+                Either::Left((backend_value, previous_fut)) => {
+                    let Some(message) = backend_value else {
+                        tracing::trace!(target: LOG_TARGET, "Frontend channel closed");
+                        break;
+                    };
+                    tracing::trace!(target: LOG_TARGET, "Received register message {:?}", message);
+
+                    self.handle_register(message).await;
+
+                    backend_event_fut = backend_event.next();
+                    rpc_responses_fut = previous_fut;
+                }
+                // Message received from rpc handler: lightclient response.
+                Either::Right((response, previous_fut)) => {
+                    // Smoldot returns `None` if the chain has been removed (which subxt does not remove).
+                    let Some(response) = response else {
+                        tracing::trace!(target: LOG_TARGET, "Smoldot RPC responses channel closed");
+                        break;
+                    };
+                    tracing::trace!(target: LOG_TARGET, "Received smoldot RPC result {:?}", response);
+
+                    self.handle_rpc_response(response).await;
+
+                    // Advance backend, save frontend.
+                    backend_event_fut = previous_fut;
+                    rpc_responses_fut = rpc_responses_event.next();
+                }
+            }
+        }
+
+        tracing::trace!(target: LOG_TARGET, "Task closed");
+    }
+}
+
+/// The RPC response from the light-client.
+/// This can either be a response of a method, or a notification from a subscription.
+#[derive(Debug, Clone)]
+enum RpcResponse {
```

Comment:
you could also use the types from jsonrpsee but I guess jsonrpsee isn't used here anyway..

Reply:
Since jsonrpsee is behind a feature flag it'd be best to avoid using it outside of where necessary anyways :)